File: paleo_workbench/catalog/gc.py
"""Conservative garbage collection for the catalog artifacts tree (P4).

``plan_gc`` classifies orphaned files with ZERO deletion; ``sweep_gc`` deletes
only the safe classes; ``cleanup_working_copies`` is the explicit working-copy
hook. Invariants:

- A reachable committed ``DataVersion`` payload is NEVER deleted: referenced
  paths are computed from the document before any sweep.
- External source files are never touched (they live outside ``.artifacts/``).
- Trashed payloads are only swept when their version record is gone (i.e. the
  trash entry has no version id anymore).
- Blobs are swept by reachability (the keep-set is every managed version's
  recorded digest — see ``dedup.py``).
- Working copies are treated as user work: only dirs whose version id does
  not exist at all are considered abandoned, and even those require the
  explicit ``cleanup_working_copies`` hook (never swept automatically).
"""

from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path

from paleo_workbench.catalog.storage import (
    STAGE_DIRS,
    blob_dir_for,
    catalog_dir_for,
    trash_dir_for,
    working_dir_for,
)
from paleo_workbench.catalog import dedup

# Orphan classes (report only; deletion policy lives in sweep_gc).
STAGE_ORPHAN = "stage_orphan"
WORKING_ORPHAN = "working_orphan"
TEMP_ORPHAN = "temp_orphan"
TRASH_ORPHAN = "trash_orphan"
BLOB_ORPHAN = "blob_orphan"
EMPTY_DIR = "empty_dir"

_TEMP_PREFIXES = (".place-", ".blob-", ".catalog.json.")
_TEMP_SUFFIX = ".tmp"


@dataclass
class GcItem:
    """One orphaned file/dir candidate."""

    kind: str
    path: Path
    size: int = 0


@dataclass
class GcReport:
    """Dry-run classification result (nothing is deleted by planning)."""

    items: list[GcItem] = field(default_factory=list)

    def by_kind(self, kind: str) -> list[GcItem]:
        return [item for item in self.items if item.kind == kind]

    def count(self, kind: str | None = None) -> int:
        if kind is None:
            return len(self.items)
        return len(self.by_kind(kind))

    def bytes_for(self, kind: str) -> int:
        return sum(item.size for item in self.by_kind(kind))

    def total_bytes(self) -> int:
        return sum(item.size for item in self.items)

    def kinds(self) -> list[str]:
        return sorted({item.kind for item in self.items})


def _is_temp_name(name: str) -> bool:
    return name.endswith(_TEMP_SUFFIX) or name.startswith(_TEMP_PREFIXES)


def _walk_files(directory: Path) -> list[Path]:
    if not directory.is_dir():
        return []
    return [p for p in directory.rglob("*") if p.is_file()]


def _temp_scan_roots(project_path: Path) -> tuple[tuple[Path, Path]]:
    """Directory roots the temp-name scan must NOT descend into.

    ``working/`` holds the only mutable copy of uncommitted user edits —
    even a temp-named working copy is user work and may only be removed via
    the explicit ``cleanup_working_copies`` hook (module invariant).
    ``trash/`` payloads are recoverable by design and are governed by
    step-3 version reachability instead.
    """
    return (working_dir_for(project_path), trash_dir_for(project_path))


def _in_roots(path: Path, roots) -> bool:
    return any(path == root or root in path.parents for root in roots)


def _plan_auto_gc(service) -> GcReport:
    """Open-time plan: temp leftovers and empty dirs only (no full-tree orphan scan)."""
    report = GcReport()
    project_path = service.project_path
    document = service.document
    referenced = {v.path for v in document.versions if v.managed}
    stage_root = catalog_dir_for(project_path).parent
    skip_roots = _temp_scan_roots(project_path)
    for path in _walk_files(stage_root):
        if not _is_temp_name(path.name):
            continue
        if _in_roots(path, skip_roots):
            continue
        try:
            rel = path.relative_to(stage_root.parent).as_posix()
        except ValueError:
            rel = ""
        if rel not in referenced:
            report.items.append(GcItem(TEMP_ORPHAN, path, _safe_size(path)))
    for directory in _empty_dirs(stage_root):
        report.items.append(GcItem(EMPTY_DIR, directory, 0))
    return report


def plan_gc(service, *, explicit: bool = True) -> GcReport:
    """Classify orphans in the artifacts tree (never deletes anything).

    ``explicit=False`` (open-time auto sweep) only classifies TEMP_ORPHAN and
    EMPTY_DIR so project open does not pay three full artifacts-tree walks.
    """
    if not explicit:
        return _plan_auto_gc(service)
    report = GcReport()
    project_path = service.project_path
    document = service.document

    # 1. Stage payloads not referenced by any managed version path.
    referenced = {v.path for v in document.versions if v.managed}
    stage_root = catalog_dir_for(project_path).parent
    for stage_name in STAGE_DIRS.values():
        for path in _walk_files(stage_root / stage_name):
            try:
                # Version paths are stored relative to the PROJECT dir and
                # include the ``<project>.artifacts/`` prefix.
                rel = path.relative_to(stage_root.parent).as_posix()
            except ValueError:
                continue
            if rel not in referenced:
                report.items.append(
                    GcItem(STAGE_ORPHAN, path, _safe_size(path))
                )

    # 2. Abandoned working copies: dirs/files whose version id is unknown.
    known_version_ids = {v.id for v in document.versions}
    working_root = working_dir_for(project_path)
    if working_root.is_dir():
        for path in _walk_files(working_root):
            try:
                version_id = path.relative_to(working_root).parts[0]
            except (ValueError, IndexError):
                version_id = ""
            if version_id not in known_version_ids:
                report.items.append(
                    GcItem(WORKING_ORPHAN, path, _safe_size(path))
                )

    # 3. Unreferenced trash payloads: trash/{version_id}/file where the
    #    version record is gone.
    trash_root = trash_dir_for(project_path)
    if trash_root.is_dir():
        for path in _walk_files(trash_root):
            try:
                version_id = path.relative_to(trash_root).parts[0]
            except (ValueError, IndexError):
                version_id = ""
            if version_id not in known_version_ids:
                report.items.append(
                    GcItem(TRASH_ORPHAN, path, _safe_size(path))
                )

    # 4. Stale temp files anywhere in the artifacts tree. A file whose name
    #    matches the temp pattern is only an orphan when NO managed version
    #    references it — an imported payload may legitimately be named
    #    ``data.tmp`` / ``.blob-x`` (review finding: auto-sweep deleted
    #    referenced payloads on open). Referenced files are NEVER classified
    #    here (same referenced set as step 1). The working/ and trash/
    #    subtrees are skipped: temp-named working copies are uncommitted user
    #    work (only ``cleanup_working_copies`` may touch them) and trash is
    #    governed by step-3 reachability (#889).
    referenced = {v.path for v in document.versions if v.managed}
    skip_roots = _temp_scan_roots(project_path)
    for path in _walk_files(stage_root):
        if not _is_temp_name(path.name):
            continue
        if _in_roots(path, skip_roots):
            continue
        try:
            rel = path.relative_to(stage_root.parent).as_posix()
        except ValueError:
            rel = ""
        if rel not in referenced:
            report.items.append(GcItem(TEMP_ORPHAN, path, _safe_size(path)))

    # 5. Unreferenced blobs (reachability GC on the content store).
    for digest in dedup.plan_blob_gc(project_path, document):
        blob_path = blob_dir_for(project_path) / digest[:2] / digest
        report.items.append(GcItem(BLOB_ORPHAN, blob_path, _safe_size(blob_path)))

    # 6. Empty directories under the payload dirs (crash leftovers).
    for directory in _empty_dirs(stage_root):
        report.items.append(GcItem(EMPTY_DIR, directory, 0))
    return report


def _safe_size(path: Path) -> int:
    try:
        return path.stat().st_size
    except OSError:
        return 0


def _empty_dirs(stage_root: Path) -> list[Path]:
    """Empty directories below the stage/working/trash trees (deepest first)."""
    empties: list[Path] = []
    for root_name in [*STAGE_DIRS.values(), "working", "trash"]:
        root = stage_root / root_name
        if not root.is_dir():
            continue
        for directory in sorted(
            (p for p in root.rglob("*") if p.is_dir()), key=lambda p: -len(p.parts)
        ):
            try:
                if not any(directory.iterdir()):
                    empties.append(directory)
            except OSError:
                continue
    return empties


# Classes the default sweep deletes (conservative). Working copies and trash
# payloads with a live version record are never swept.
_AUTO_SWEEPABLE = {TEMP_ORPHAN, EMPTY_DIR}
_EXPLICIT_SWEEPABLE = {
    STAGE_ORPHAN,
    TEMP_ORPHAN,
    TRASH_ORPHAN,
    BLOB_ORPHAN,
    EMPTY_DIR,
}


def sweep_gc(service, *, dry_run: bool = True, explicit: bool = False) -> GcReport:
    """Plan (dry_run=True) or perform a GC sweep.

    With ``explicit=False`` (the conservative sweep used on open) only stale
    temp files and empty directories are removed. With ``explicit=True`` the
    full safe set is swept: stage orphans, trash orphans and unreferenced
    blobs too. Working copies are never swept here (see
    :func:`cleanup_working_copies`).
    """
    report = plan_gc(service, explicit=explicit)
    sweepable = _EXPLICIT_SWEEPABLE if explicit else _AUTO_SWEEPABLE
    removed: list[GcItem] = []
    for item in report.items:
        if item.kind not in sweepable:
            continue
        if dry_run:
            removed.append(item)
            continue
        try:
            if item.path.is_dir():
                item.path.rmdir()
            else:
                item.path.unlink()
            removed.append(item)
        except PermissionError:
            # Read-only payload (blobs are content-addressed and immutable
            # by contract): Windows refuses to unlink read-only files —
            # clear the bit and retry once.
            try:
                item.path.chmod(item.path.stat().st_mode | stat.S_IWUSR)
                if item.path.is_dir():
                    item.path.rmdir()
                else:
                    item.path.unlink()
                removed.append(item)
            except OSError:
                continue
        except OSError:
            continue
    return GcReport(removed)


def cleanup_working_copies(service) -> GcReport:
    """Remove abandoned working-copy payloads (explicit user action).

    A working copy is abandoned when its directory's version id does not exist
    in the document at all (the version was purged, or the copy was created
    and the version never committed). Live versions' working copies are never
    touched — they may hold uncommitted user edits.
    """
    report = GcReport()
    known_version_ids = {v.id for v in service.document.versions}
    working_root = working_dir_for(service.project_path)
    if not working_root.is_dir():
        return report
    for path in _walk_files(working_root):
        try:
            version_id = path.relative_to(working_root).parts[0]
        except (ValueError, IndexError):
            version_id = ""
        if version_id not in known_version_ids:
            try:
                path.unlink()
                report.items.append(GcItem(WORKING_ORPHAN, path, 0))
            except OSError:
                continue
    # Prune now-empty working subdirs (deepest first).
    for directory in sorted(
        (p for p in working_root.rglob("*") if p.is_dir()), key=lambda p: -len(p.parts)
    ):
        try:
            if not any(directory.iterdir()):
                directory.rmdir()
        except OSError:
            continue
    return report

File: libs/catalog/src/gc.cpp
#include "pwb/catalog/gc.hpp"

#include "pwb/catalog/dedup.hpp"
#include "pwb/project/paths.hpp"

#include <algorithm>
#include <chrono>
#include <ctime>
#include <set>

namespace pwb::catalog {

namespace {

namespace fs = std::filesystem;
using pwb::project::artifact_dir_for;
using pwb::project::project_dir_for;

using pwb::catalog::kStageDirs;  // storage.py STAGE_DIRS: on-disk dir names
const char* const kTempScanSkipRoots[] = {"working", "trash"};

bool is_temp_name(const std::string& name) {
    static const char* const kPrefixes[] = {".place-", ".blob-",
                                            ".catalog.json."};
    if (name.size() >= 4 && name.compare(name.size() - 4, 4, ".tmp") == 0) {
        return true;
    }
    for (const char* prefix : kPrefixes) {
        if (name.rfind(prefix, 0) == 0) return true;
    }
    return false;
}

bool is_leased(const std::string& rel, const std::set<std::string>& leased) {
    for (const std::string& target : leased) {
        if (rel == target
            || (rel.size() > target.size() && rel.compare(0, target.size(), target) == 0
                && rel[target.size()] == '/')) {
            return true;
        }
    }
    return false;
}

bool is_under(const fs::path& path, const fs::path& base) {
    const std::string base_text = base.generic_string();
    const std::string path_text = path.generic_string();
    if (path_text == base_text) return true;
    if (base_text.empty()) return false;
    return path_text.rfind(base_text + "/", 0) == 0;
}

fs::path artifacts_root(const GcContext& context) {
    return artifact_dir_for(context.project_path);
}

fs::path project_root(const GcContext& context) {
    return project_dir_for(context.project_path);
}

// Python `_walk_files`: every regular file below *directory*.
std::vector<fs::path> walk_files(const fs::path& directory) {
    std::vector<fs::path> files;
    std::error_code ec;
    if (!fs::is_directory(directory, ec)) return files;
    for (const fs::directory_entry& entry :
         fs::recursive_directory_iterator(
             directory, fs::directory_options::skip_permission_denied, ec)) {
        std::error_code file_ec;
        if (entry.is_regular_file(file_ec) && !file_ec) {
            files.push_back(entry.path());
        }
    }
    return files;
}

std::string rel_or_empty(const fs::path& project_dir, const fs::path& path) {
    std::error_code ec;
    const fs::path rel = fs::relative(path, project_dir, ec);
    if (ec) return "";
    return pwb::project::path_to_u8(rel);
}

std::int64_t safe_size(const fs::path& path) {
    std::error_code ec;
    const auto size = fs::file_size(path, ec);
    return ec ? 0 : static_cast<std::int64_t>(size);
}

// Empty directories below *root*, deepest first (Python's per-root rglob
// with the -len(parts) ordering).
std::vector<fs::path> empty_dirs_below(const fs::path& root) {
    std::vector<fs::path> empties;
    std::error_code ec;
    if (!fs::is_directory(root, ec)) return empties;
    std::vector<fs::path> dirs;
    for (const fs::directory_entry& entry :
         fs::recursive_directory_iterator(
             root, fs::directory_options::skip_permission_denied, ec)) {
        std::error_code dir_ec;
        if (entry.is_directory(dir_ec) && !dir_ec) {
            dirs.push_back(entry.path());
        }
    }
    std::sort(dirs.begin(), dirs.end(), [](const fs::path& a,
                                           const fs::path& b) {
        const auto da = std::distance(a.begin(), a.end());
        const auto db = std::distance(b.begin(), b.end());
        if (da != db) return da > db;  // deepest first
        return a.generic_string() < b.generic_string();
    });
    for (const fs::path& dir : dirs) {
        std::error_code empty_ec;
        if (fs::is_empty(dir, empty_ec) && !empty_ec) {
            empties.push_back(dir);
        }
    }
    return empties;
}

// Empty directories below the stage/working/trash trees (gc.py `_empty_dirs`).
std::vector<fs::path> empty_dirs(const fs::path& stage_root) {
    std::vector<fs::path> empties;
    for (const char* root_name : kStageDirs) {
        auto sub = empty_dirs_below(stage_root / root_name);
        empties.insert(empties.end(), sub.begin(), sub.end());
    }
    auto working = empty_dirs_below(stage_root / "working");
    empties.insert(empties.end(), working.begin(), working.end());
    auto trash = empty_dirs_below(stage_root / "trash");
    empties.insert(empties.end(), trash.begin(), trash.end());
    return empties;
}

std::set<std::string> referenced_paths(const CatalogDocument& document) {
    std::set<std::string> paths;
    for (const DataVersion& version : document.versions) {
        if (version.managed) paths.insert(version.path);
    }
    return paths;
}

std::set<std::string> known_version_ids(const CatalogDocument& document) {
    std::set<std::string> ids;
    for (const DataVersion& version : document.versions) {
        ids.insert(version.id.str());
    }
    return ids;
}

}  // namespace

std::int64_t GcReport::count(std::optional<std::string_view> kind) const {
    if (!kind.has_value()) {
        return static_cast<std::int64_t>(items.size());
    }
    std::int64_t total = 0;
    for (const GcItem& item : items) {
        if (item.kind == *kind) total += 1;
    }
    return total;
}

std::int64_t GcReport::bytes_for(std::string_view kind) const {
    std::int64_t total = 0;
    for (const GcItem& item : items) {
        if (item.kind == kind) total += item.size;
    }
    return total;
}

std::int64_t GcReport::total_bytes() const {
    std::int64_t total = 0;
    for (const GcItem& item : items) total += item.size;
    return total;
}

std::set<std::string> active_staging_targets(const CatalogDocument& document,
                                             const std::string& cutoff_iso) {
    std::set<std::string> targets;
    for (const StagingLease& lease : document.staging_leases) {
        if (lease.heartbeat_at > cutoff_iso) targets.insert(lease.target);
    }
    return targets;
}

// CONV-31b (A5, findings §F-2 R-leak fix): the live read Python actually
// performs (db.py 1289-1308 queries sqlite; it never consults the loaded
// document). A lease acquired after the caller's document snapshot loaded
// — exactly the lease guarding in-flight registration bytes — is invisible
// to the snapshot form above, so plan/sweep must use this one.
std::set<std::string> active_staging_targets_live(Database& database,
                                                  const std::string& cutoff_iso) {
    std::set<std::string> targets;
    if (!database.is_open()) return targets;
    if (!database.table_exists("staging_leases")) return targets;
    Statement statement = database.prepare(
        "SELECT DISTINCT target FROM staging_leases WHERE heartbeat_at > ?");
    if (!statement.is_valid()) return targets;
    statement.bind(1, cutoff_iso);
    while (statement.step()) {
        targets.insert(statement.text(0));
    }
    return targets;
}

std::set<std::string> active_staging_targets_live(
    const fs::path& project_path, const std::string& cutoff_iso) {
    auto opened = Database::open(
        pwb::project::catalog_sqlite_for(project_path),
        SqliteOpenMode::ReadOnly);
    if (!opened.is_ok()) {
        return {};  // missing/unreadable store: no leases (swallow parity)
    }
    return active_staging_targets_live(opened.value(), cutoff_iso);
}

std::string default_lease_cutoff() {
    const auto now = std::chrono::system_clock::now()
        - std::chrono::seconds(3600);  // db.py STAGING_LEASE_TTL_SECONDS
    const std::time_t time = std::chrono::system_clock::to_time_t(now);
    std::tm local{};
#if !defined(_WIN32)
    localtime_r(&time, &local);
#else
    localtime_s(&local, &time);
#endif
    char buffer[32];
    const std::size_t written =
        std::strftime(buffer, sizeof(buffer), "%Y-%m-%dT%H:%M:%S", &local);
    return std::string(buffer, written);
}

GcReport plan_gc(const GcContext& context, bool explicit_plan) {
    GcReport report;
    const CatalogDocument& document = *context.document;
    const fs::path dir = project_root(context);
    const fs::path stage_root = artifacts_root(context);
    static_assert(std::size(pwb::catalog::kStageDirs) == 4);
    const std::set<std::string> referenced = referenced_paths(document);
    // CONV-31b: live sqlite lease read (the document snapshot cannot see a
    // lease acquired after the caller loaded it — findings §F-2).
    const std::set<std::string> leased =
        active_staging_targets_live(context.project_path,
                                    default_lease_cutoff());

    auto classify_temp_and_empty = [&](GcReport& into) {
        for (const fs::path& path : walk_files(stage_root)) {
            if (!is_temp_name(path.filename().generic_string())) continue;
            bool skipped = false;
            for (const char* skip : kTempScanSkipRoots) {
                // working/ and trash/ are skipped wholesale for temp names:
                // a temp-named working copy is uncommitted user work and
                // trash is reachability-governed (#889).
                if (is_under(path, stage_root / skip)) {
                    skipped = true;
                    break;
                }
            }
            if (skipped) continue;
            const std::string rel = rel_or_empty(dir, path);
            if (referenced.count(rel) != 0) continue;  // #889: referenced wins
            if (is_leased(rel, leased)) continue;
            into.items.push_back(
                {kGcTempOrphan, rel, safe_size(path)});
        }
        for (const fs::path& directory : empty_dirs(stage_root)) {
            const std::string rel = rel_or_empty(dir, directory);
            if (is_leased(rel, leased)) continue;
            into.items.push_back({kGcEmptyDir, rel, 0});
        }
    };

    if (!explicit_plan) {
        classify_temp_and_empty(report);
        return report;
    }

    // 1. Stage payloads not referenced by any managed version path.
    for (const char* stage_name : kStageDirs) {
        for (const fs::path& path : walk_files(stage_root / stage_name)) {
            const std::string rel = rel_or_empty(dir, path);
            if (rel.empty()) continue;
            if (referenced.count(rel) != 0) continue;
            if (is_leased(rel, leased)) continue;
            report.items.push_back({kGcStageOrphan, rel, safe_size(path)});
        }
    }

    // 2. Abandoned working copies; 3. unreferenced trash payloads.
    const std::set<std::string> version_ids = known_version_ids(document);
    auto classify_first_segment = [&](const char* root_name, const char* kind) {
        const fs::path root = stage_root / root_name;
        std::error_code ec;
        if (!fs::is_directory(root, ec)) return;
        for (const fs::path& path : walk_files(root)) {
            const fs::path rel = fs::relative(path, root, ec);
            if (ec || rel.empty()) continue;
            const std::string first = rel.begin()->generic_string();
            if (version_ids.count(first) != 0) continue;
            report.items.push_back(
                {kind, rel_or_empty(dir, path), safe_size(path)});
        }
    };
    classify_first_segment("working", kGcWorkingOrphan);
    classify_first_segment("trash", kGcTrashOrphan);

    // 4. Stale temp files anywhere (working/ + trash/ skipped, referenced
    // payloads never classified); 6. empty dirs.
    classify_temp_and_empty(report);

    // 5. Unreferenced blobs (reachability GC on the content store).
    for (const std::string& digest :
         plan_blob_gc(context.project_path, document)) {
        const fs::path blob = blob_path_for(context.project_path, digest);
        const std::string rel = rel_or_empty(dir, blob);
        if (is_leased(rel, leased)) continue;
        report.items.push_back({kGcBlobOrphan, rel, safe_size(blob)});
    }
    return report;
}

GcReport sweep_gc(const GcContext& context, bool dry_run,
                  bool explicit_sweep, const GcReport* precomputed) {
    static const char* const kAutoSweepable[] = {kGcTempOrphan, kGcEmptyDir};
    static const char* const kExplicitSweepable[] = {
        kGcStageOrphan, kGcTempOrphan, kGcTrashOrphan, kGcBlobOrphan,
        kGcEmptyDir};
    const GcReport source =
        precomputed != nullptr ? *precomputed
                               : plan_gc(context, explicit_sweep);
    auto sweepable = [&](const std::string& kind) {
        if (explicit_sweep) {
            for (const char* k : kExplicitSweepable) {
                if (kind == k) return true;
            }
        } else {
            for (const char* k : kAutoSweepable) {
                if (kind == k) return true;
            }
        }
        return false;
    };

    GcReport candidates;
    for (const GcItem& item : source.items) {
        if (sweepable(item.kind)) candidates.items.push_back(item);
    }
    if (dry_run) return candidates;

    const fs::path dir = project_root(context);
    const CatalogDocument& document = *context.document;
    const std::set<std::string> referenced = referenced_paths(document);
    // CONV-31b: re-validation also uses the live lease read (db.py 1289-1308
    // parity; the snapshot form misses in-flight leases — findings §F-2).
    const std::set<std::string> leased =
        active_staging_targets_live(context.project_path,
                                    default_lease_cutoff());

    GcReport removed;
    for (const GcItem& item : candidates.items) {
        // Re-validate the point-in-time plan: registered or staged since the
        // plan → keep it.
        if (referenced.count(item.rel_path) != 0) continue;
        if (is_leased(item.rel_path, leased)) continue;
        const fs::path absolute = dir / pwb::project::path_from_u8(item.rel_path);
        std::error_code ec;
        if (!fs::remove(absolute, ec)) {
            // Read-only accident guard (Windows refuses to unlink; POSIX
            // ignores the bit): add +u+w and retry once, then stay
            // conservative (gc.py chmod(mode | S_IWUSR) parity).
            std::error_code retry_ec;
            fs::permissions(absolute, fs::perms::owner_write,
                            fs::perm_options::add, retry_ec);
            if (!fs::remove(absolute, retry_ec) || retry_ec) {
                continue;
            }
        }
        removed.items.push_back(item);
    }
    return removed;
}

GcReport cleanup_working_copies(const GcContext& context) {
    GcReport report;
    const CatalogDocument& document = *context.document;
    const std::set<std::string> version_ids = known_version_ids(document);
    const fs::path dir = project_root(context);
    const fs::path working_root =
        artifact_dir_for(context.project_path) / "working";
    std::error_code ec;
    if (!fs::is_directory(working_root, ec)) return report;
    for (const fs::path& path : walk_files(working_root)) {
        const fs::path rel = fs::relative(path, working_root, ec);
        if (ec || rel.empty()) continue;
        const std::string first = rel.begin()->generic_string();
        if (version_ids.count(first) != 0) continue;
        if (!fs::remove(path, ec) || ec) continue;
        report.items.push_back({kGcWorkingOrphan, rel_or_empty(dir, path), 0});
    }
    // Prune now-empty working subdirs (deepest first).
    for (const fs::path& directory : empty_dirs_below(working_root)) {
        std::error_code rm_ec;
        fs::remove(directory, rm_ec);  // rmdir semantics: fails if non-empty
    }
    return report;
}

}  // namespace pwb::catalog
